internal/controller/flow_cilium.go

```diff
@@ -0,0 +1,208 @@
+	if err != nil {
+		return nil, fmt.Errorf("failed to connect to Cilium Hubble Relay: %w", err)
+	}
+	hubbleClient := observer.NewObserverClient(conn)
+	return &Collector{logger: logger, client: hubbleClient}, nil
+}
+
+// convertLayer4 function converts a slice of flow.Layer4 objects to a slice of pb.Layer4 objects.
+func convertLayer4(l4 *flow.Layer4) *pb.Layer4 {
+	layer4 := &pb.Layer4{}
+	if l4 == nil {
+		return layer4
+	}
+	switch protocol := l4.Protocol.(type) {
+	case *flow.Layer4_TCP:
+		layer4.Protocol = &pb.Layer4_Tcp{
+			Tcp: &pb.TCP{
+				SourcePort:      protocol.TCP.SourcePort,
+				DestinationPort: protocol.TCP.DestinationPort,
+			},
+		}
+	case *flow.Layer4_UDP:
+		layer4.Protocol = &pb.Layer4_Udp{
+			Udp: &pb.UDP{
+				SourcePort:      protocol.UDP.SourcePort,
+				DestinationPort: protocol.UDP.DestinationPort,
+			},
+		}
+	case *flow.Layer4_ICMPv4:
+		layer4.Protocol = &pb.Layer4_Icmpv4{
+			Icmpv4: &pb.ICMPv4{
+				Type: protocol.ICMPv4.Type,
+				Code: protocol.ICMPv4.Code,
+			},
+		}
+	case *flow.Layer4_ICMPv6:
+		layer4.Protocol = &pb.Layer4_Icmpv6{
+			Icmpv6: &pb.ICMPv6{
+				Type: protocol.ICMPv6.Type,
+				Code: protocol.ICMPv6.Code,
+			},
+		}
+	case *flow.Layer4_SCTP:
+		layer4.Protocol = &pb.Layer4_Sctp{
+			Sctp: &pb.SCTP{
+				SourcePort:      protocol.SCTP.SourcePort,
+				DestinationPort: protocol.SCTP.DestinationPort,
+			},
+		}
+	default:
+	}
+
+	return layer4
+}
+
+// convertCiliumWorkflows function converts a slice of flow.Workload objects to a slice of pb.Workload objects.
+func convertCiliumWorkflows(workloads []*flow.Workload) []*pb.Workload {
+	protoWorkloads := []*pb.Workload{}
+	for _, workload := range workloads {
+		protoWorkload := &pb.Workload{
+			Name: workload.GetName(),
+			Kind: workload.GetKind(),
+		}
+		protoWorkloads = append(protoWorkloads, protoWorkload)
+	}
+	return protoWorkloads
+}
+
+// convertCiliumPolicies function converts a slice of flow.Policy objects to a slice of pb.Policy objects.
+func convertCiliumPolicies(policies []*flow.Policy) []*pb.Policy {
+	protoPolicies := []*pb.Policy{}
+	for _, policy := range policies {
+		protoPolicy := &pb.Policy{
+			Name:      policy.GetName(),
+			Namespace: policy.GetNamespace(),
+			Labels:    policy.GetLabels(),
+			Revision:  policy.GetRevision(),
+		}
+		protoPolicies = append(protoPolicies, protoPolicy)
+	}
+	return protoPolicies
+}
+
+// readFlows loops continously while using the observerClient to make gRPC calls to hubble-relay and grabs the last x amount of flows.
+// It sends converts those flows to ciliumFlows then sends them over the open stream.
+func (fm *Collector) readFlows(ctx context.Context, sm streamManager) error {
+	req := &observer.GetFlowsRequest{
+		Number: flowCount,
+		Follow: true,
+	}
+	observerClient := fm.client
+	stream, err := observerClient.GetFlows(ctx, req)
+	if err != nil {
+		fm.logger.Errorw("Error getting network flows", "error", err)
+		return err
+	}
```

Comment:
You must close this stream when returning, in a `defer`.
```suggestion
	stream, err := observerClient.GetFlows(ctx, req)
	if err != nil {
		fm.logger.Errorw("Error getting network flows", "error", err)
		return err
	}
	defer ...
```